song2txt/txt_parser/metadata.py

```diff
@@ -0,0 +1,155 @@
+        self.edition: Optional[str] = None
+        self.creator: Optional[str] = None
+        self.language: Optional[str] = None
+        self.cover: Optional[PathLike] = None
+        self.vocals: Optional[PathLike] = None
+        self.video: Optional[PathLike] = None
+        self.background: Optional[PathLike] = None
+        self.relative: Optional[bool] = None
+        self.gap: Optional[float] = None
+        self.video_gap: Optional[float] = None
+        self.preview_start: Optional[float] = None
+        self.name_singer_1: Optional[str] = None
+        self.name_singer_2: Optional[str] = None
+        self.extra = {}
+
+    def __str__(self) -> str:
+        model = '#{}:{}\n'
+
+        text = model.format('TITLE', self.title)
+        text += model.format('ARTIST', self.artist)
+        text += model.format('MP3', self.mp3)
+        text += model.format('BPM', self._float_str(self.bpm))
+        if self.genre is not None:
+            text += model.format('GENRE', self.genre)
+        if self.year is not None:
+            text += model.format('YEAR', self.year)
+        if self.edition is not None:
+            text += model.format('EDITION', self.edition)
+        if self.creator is not None:
+            text += model.format('CREATOR', self.creator)
+        if self.language is not None:
+            text += model.format('LANGUAGE', self.language)
+        if self.cover is not None:
+            text += model.format('COVER', self.cover)
+        if self.vocals is not None:
+            text += model.format('VOCALS', self.vocals)
+        if self.video is not None:
+            text += model.format('VIDEO', self.video)
+        if self.background is not None:
+            text += model.format('BACKGROUND', self.background)
+        if self.relative is not None:
+            text += model.format('RELATIVE', 'YES' if self.relative else 'NO')
+        if self.gap is not None:
+            text += model.format('GAP', self._float_str(self.gap))
+        if self.video_gap is not None:
+            text += model.format('VIDEOGAP', self._float_str(self.video_gap))
+        if self.preview_start is not None:
+            text += model.format('PREVIEWSTART',
+                                 self._float_str(self.preview_start))
+        if self.name_singer_1 is not None:
+            text += model.format('DUETSINGERP1', self.name_singer_1)
+        if self.name_singer_2 is not None:
+            text += model.format('DUETSINGERP2', self.name_singer_2)
+        for key, value in self.extra.items():
+            text += model.format(key, value)
+
+        return text
+
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, MetaData) and self.__dict__ == other.__dict__
+
+    @staticmethod
+    def _float_str(num: float) -> str:
+        if num.is_integer():
+            return str(int(num))
+        else:
+            return str(num)
+
+    def get_bpm(self):
+        return self._bpm
+
+    def set_bpm(self, bpm):
```

Comment:
Uf, getters y setters. ¿Es un objeto estático? ¿Hay algún requisito o alguna historia de usuario que pida que se cambien? Si no lo hay, usa una `@dataclass`

Reply:
Creo que es útil que se puedan cambiar ya que no se usan solo para consultar los valores original del txt, si no para generar un txt con nuevos valores, y los objetos tienen la funcionalidad para generar la representación escrita. En #37 se comenta que la idea es extraer los datos, modificarlos y generar el nuevo txt tanto para #7 como para #8.

¿Hay algún problema con definir getters y setters? Ten en cuenta que no se usarían como funciones, el acceso / asignación a la variable se sigue haciendo con metadata.bpm. La segunda respuesta explica bastante bien el razonamiento https://stackoverflow.com/questions/2627002/whats-the-pythonic-way-to-use-getters-and-setters , la sintaxis que he usado es un poco diferente pero me parecía un poco más clara y es la que viene aquí https://docs.python.org/3/library/functions.html#property

La idea es lanzar una excepción si se asignan valores imposibles, y los getters en principio no sirven para nada pero estoy obligado a definirlos para cambiar el comportamiento al asignar.

Reply:
Otro argumento a favor de que sea una entidad y no un objeto valor: poder usar la biblioteca en editores de canciones. Cada vez que muevas ligeramente una nota de sitio, cambies su tono, letra, etc... crear otro objeto inmutable completo es engorroso y poco eficiente.

Reply:
No se trata de que tú creas que es útil o no, sino de lo que añada valor al cliente y se corresponda con la aplicación; para eso tienes que seguir una metodología que mapee objetos del dominio del problema al de la aplicación. En ausencia de esa metodología, usar algo "por si" no es nunca una buena idea

Reply:
Quiero decir, "metadatos" no parece algo que vaya a variar durante la vida de un objeto. Puede haber datos que vayas a convertir en metadatos, pero ya está.

Reply:
Le he dado una vuelta, considero que con bastante detenimiento, en #45. Si se te ocurre cualquier cosa que debería tener en cuenta al respecto soy todo oídos.